src/canister_tests/src/certificate_validation.rs

```diff
@@ -0,0 +1,147 @@
+// adapted from https://github.com/dfinity/icx-proxy/blob/0cd1a22f717b56ac550a3554a25a845878bfb4e8/src/main.rs#L611
+// TODO: certificate validation should be its own library
+
+use crate::certificate_validation::ValidationError::{
+    AssetHashMismatch, AssetPathLookupFailed, MalformedCertificate, TreeHashCertifiedDataMismatch,
+    WitnessLookupFailed,
+};
+use candid::types::ic_types::hash_tree::LookupResult;
+use flate2::read::GzDecoder;
+use ic_agent::agent::http_transport::ReqwestHttpReplicaV2Transport;
+use ic_agent::hash_tree::HashTree;
+use ic_agent::{lookup_value, Agent, AgentError};
+use ic_crypto_utils_threshold_sig_der::public_key_to_der;
+use ic_state_machine_tests::{CanisterId, ThresholdSigPublicKey};
+use regex::Regex;
+use serde::Deserialize;
+use sha2::{Digest, Sha256};
+use std::io::Read;
+
+// The limit of a buffer we should decompress ~10mb.
+const MAX_CHUNK_SIZE_TO_DECOMPRESS: usize = 1024;
+const MAX_CHUNKS_TO_DECOMPRESS: u64 = 10_240;
+
+const CERTIFICATE_REGEX: &'static str = "^certificate=:([^:]*):, tree=:([^:]*):$";
```

Comment:
```suggestion
const CERTIFICATE_REGEX: &'static str = "^certificate=:([^:]*):, tree=:([^:]*):$";
```

nit: must there be exactly one whitespace between the values or can there be an arbitrary number?